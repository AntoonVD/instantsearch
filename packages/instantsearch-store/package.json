{
  "name": "instantsearch-store",
  "version": "0.0.0",
  "files": [
    "dist"
  ],
  "main": "dist/instantsearch-store.common.js",
  "module": "dist/instantsearch-store.esm.js",
  "jsnext:main": "dist/instantsearch-store.esm.js",
  "unpkg": "dist/instantsearch-store.js",
  "scripts": {
    "build": "yarn build:umd && yarn build:es && yarn build:cjs",
    "build:umd": "rollup -c ../../build/rollup.umd.config.js",
    "build:es": "rollup -c ../../build/rollup.es.config.js",
    "build:cjs": "rollup -c ../../build/rollup.cjs.config.js"
  },
  "dependencies": {
    "algoliasearch": "^3.18.1",
    "algoliasearch-helper": "^2.14.0"
  },
  "peerDependencies": {
    "vue": "^2.2.2"
  }
}
