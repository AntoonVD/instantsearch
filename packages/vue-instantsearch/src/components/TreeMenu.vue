<script>
import { FACET_TREE } from '../store';
import algoliaComponent from '../component';

export default {
  mixins: [algoliaComponent],
  props: {
    attribute: {
      type: String,
      default: 'tree-menu',
    },
    attributes: {
      type: Array,
      required: true,
    },
    separator: {
      type: String,
      default: ' > ',
    },
    limit: {
      type: Number,
      default: 10,
    },
    sortBy: {
      default() {
        return ['name:asc'];
      },
    },
  },
  data() {
    return {
      widgetName: 'ais-tree-menu',
    };
  },
  created() {
    this.searchStore.addFacet(
      {
        name: this.attribute,
        attributes: this.attributes,
        separator: this.separator,
      },
      FACET_TREE
    );
  },
  destroyed() {
    this.searchStore.stop();
    this.searchStore.removeFacet(this.attribute);
    this.searchStore.start();
  },
  computed: {
    facetValues() {
      const values = this.searchStore.getFacetValues(
        this.attribute,
        this.sortBy
      );

      return values.data || [];
    },
    show() {
      return this.facetValues.length > 0;
    },
  },
  methods: {
    toggleRefinement(value) {
      return this.searchStore.toggleFacetRefinement(this.attribute, value.path);
    },
    _renderList(h, facetValues, isRoot = true) {
      const listItems = [];
      for (const facet of facetValues) {
        const listItemLabel = [];

        if (this.$scopedSlots.default) {
          listItemLabel.push(
            this.$scopedSlots.default({
              value: facet.name,
              count: facet.count,
              active: facet.isRefined,
            })
          );
        } else {
          listItemLabel.push(
            h(
              'span',
              {
                class: this.suit('value'),
              },
              facet.name
            ),
            h(
              'span',
              {
                class: this.suit('count'),
              },
              facet.count
            )
          );
        }

        const listItemChildren = [
          h(
            'a',
            {
              domProps: {
                href: '#',
              },
              on: {
                click: event => {
                  event.preventDefault();
                  this.toggleRefinement(facet);
                },
              },
            },
            listItemLabel
          ),
        ];

        if (facet.isRefined && facet.data && facet.data.length > 0) {
          listItemChildren.push(this._renderList(h, facet.data, false));
        }

        listItems.push(
          h(
            'li',
            {
              class: [
                this.suit('item'),
                facet.isRefined ? this.suit('item', 'active') : '',
              ],
            },
            listItemChildren
          )
        );
      }

      return h(
        'ul',
        {
          class: isRoot ? this.suit('list') : '',
        },
        listItems
      );
    },
  },
  render(h) {
    if (this.show === false) {
      return undefined;
    }

    const children = [];

    if (this.$slots.header) {
      children.push(this.$slots.header);
    }

    children.push(this._renderList(h, this.facetValues));

    if (this.$slots.footer) {
      children.push(this.$slots.footer);
    }

    return h(
      'div',
      {
        class: this.suit(),
      },
      children
    );
  },
};</script>
