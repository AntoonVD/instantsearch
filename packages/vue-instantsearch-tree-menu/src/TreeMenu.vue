<template>
  <div :class="bem()" v-if="show">

    <slot name="header"></slot>

    <ul v-if="facetValues.length > 0">
      <li v-for="facet in facetValues" :class="[bem('item'), facet.isRefined ? bem('item', 'active') : '']" >

        <a href="#" @click.prevent="toggleRefinement(facet)">
          <slot :value="facet.name" :count="facet.count" :active="facet.isRefined">
            <span :class="bem('item', 'value')">{{ facet.name }}</span>
            <span :class="bem('item', 'count')">{{ facet.count }}</span>
          </slot>
        </a>

        <template v-if="facet.isRefined && facet.data.length > 0">
          <ul>
            <li v-for="subfacet in facet.data" :class="[bem('item'), subfacet.isRefined ? bem('item', 'active') : '']">
              <a href="#" @click.prevent="toggleRefinement(subfacet)">
                <slot :value="subfacet.name" :count="subfacet.count" :active="subfacet.isRefined">
                  <span :class="bem('item', 'value')">{{ subfacet.name }}</span>
                  <span :class="bem('item', 'count')">{{ subfacet.count }}</span>
                </slot>
              </a>
            </li>
          </ul>
        </template>

      </li>
    </ul>

    <slot name="footer"></slot>

  </div>
</template>

<script>
  import {FACET_TREE} from 'instantsearch-store'
  import algoliaComponent from 'vue-instantsearch-component'

  export default {
    mixins: [algoliaComponent],
    props: {
      attribute: {
        type: String,
        required: false,
        default: 'tree'
      },
      attributes: {
        type: Array,
        required: true
      },
      separator: {
        type: String,
        required: false,
        default: ' > '
      },
      limit: {
        type: Number,
        default: 10
      },
      sortBy: {
        default () {
          return ['name:asc']
        }
      }
    },
    data () {
      return {
        blockClassName: 'ais-tree-menu'
      }
    },
    mounted () {
      this.searchStore.addFacet({
        name: this.attribute,
        attributes: this.attributes,
        separator: this.separator
      }, FACET_TREE)
    },
    destroyed () {
      this.searchStore.removeFacet(this.attribute)
    },
    computed: {
      facetValues () {
        let values = this.searchStore.getFacetValues(this.attribute, this.sortBy)

        return values.data || []
      },
      show () {
        return this.facetValues.length > 0
      }
    },
    methods: {
      toggleRefinement (value) {
        return this.searchStore.toggleFacetRefinement(this.attribute, value.path)
      }
    }
  }
</script>
