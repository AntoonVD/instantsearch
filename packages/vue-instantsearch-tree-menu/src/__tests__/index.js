import { FACET_TREE } from 'instantsearch-store';
import TreeMenu from 'vue-instantsearch-tree-menu';
import Vue from 'vue';

const getFacetValues = jest.fn(attributeName => {
  return {
    name: attributeName,
    count: null,
    isRefined: true,
    path: null,
    data: [
      {
        count: 319,
        data: null,
        isRefined: false,
        name: 'Bathroom',
        path: 'Bathroom',
      },
      {
        count: 200,
        data: [
          {
            count: 43,
            data: null,
            isRefined: false,
            name: 'Bakeware',
            path: 'Cooking > Bakeware',
          },
        ],
        isRefined: true,
        name: 'Cooking',
        path: 'Cooking',
      },
    ],
  };
});
const addFacet = jest.fn();
const searchStore = {
  getFacetValues,
  addFacet,
};

test('renders proper HTML', () => {
  const Component = Vue.extend(TreeMenu);
  const vm = new Component({
    propsData: {
      attributes: ['category.lvl1', 'category.lvl2'],
      searchStore,
    },
  });
  vm.$mount();

  expect(vm.$el.outerHTML).toMatchSnapshot();
});

test('should add a tree facet to the store when mounted', () => {
  const Component = Vue.extend(TreeMenu);
  const addFacetMock = jest.fn();
  const store = Object.assign({}, searchStore, { addFacet: addFacetMock });
  const vm = new Component({
    propsData: {
      attributes: ['category.lvl1', 'category.lvl2'],
      searchStore: store,
    },
  });

  expect(addFacetMock).not.toBeCalled();
  vm.$mount();
  expect(addFacetMock).toBeCalledWith(
    {
      name: 'tree-menu',
      attributes: ['category.lvl1', 'category.lvl2'],
      separator: ' > ',
    },
    FACET_TREE
  );
});

test('should remove the tree facet from the store when destroyed', () => {
  const Component = Vue.extend(TreeMenu);
  const removeFacetMock = jest.fn();
  const store = Object.assign({}, searchStore, {
    removeFacet: removeFacetMock,
  });

  const vm = new Component({
    propsData: {
      attributes: ['category.lvl1', 'category.lvl2'],
      searchStore: store,
    },
  });

  vm.$mount();

  expect(removeFacetMock).not.toBeCalled();

  vm.$destroy();

  expect(removeFacetMock).toBeCalledWith('tree-menu');
});
