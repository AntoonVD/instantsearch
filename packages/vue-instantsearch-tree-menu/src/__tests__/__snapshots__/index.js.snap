// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders proper HTML 1`] = `

<div class="ais-tree-menu">
  <ul class="ais-tree-menu__list">
    <li class="ais-tree-menu__item">
      <a href="#">
        <span class="ais-tree-menu__value">
          Bathroom
        </span>
        <span class="ais-tree-menu__count">
          319
        </span>
      </a>
    </li>
    <li class="ais-tree-menu__item ais-tree-menu__item ais-tree-menu__item--active">
      <a href="#">
        <span class="ais-tree-menu__value">
          Cooking
        </span>
        <span class="ais-tree-menu__count">
          200
        </span>
      </a>
      <ul class>
        <li class="ais-tree-menu__item">
          <a href="#">
            <span class="ais-tree-menu__value">
              Bakeware
            </span>
            <span class="ais-tree-menu__count">
              43
            </span>
          </a>
        </li>
      </ul>
    </li>
  </ul>
</div>

`;
